### PR TITLE
Add share target

### PR DESCRIPTION
**Pull request by jakearchibald**
I'm happy with this now, although I'm happy to block on #353.

### CODE REVIEW

File: src/lib/sw-bridge.ts [hunk 1]
```diff
@@ -40,6 +40,23 @@ async function updateReady(reg: ServiceWorkerRegistration): Promise<void> {
...
+      if (event.data.action !== 'load-image') return;
+      resolve(event.data.file);
+      navigator.serviceWorker.removeEventListener('message', onmessage);
+    };
+
+    navigator.serviceWorker.addEventListener('message', onmessage);
+
+    // This message is picked up by the service worker - it's how it knows we're ready to receive
+    // the file.
+    new BroadcastChannel('share-ready').postMessage('share-ready');
```
Comment (surma) [inline src/lib/sw-bridge.ts]: What about iOS?

Comment (jakearchibald) [reply src/lib/sw-bridge.ts]: Ah fuck

Comment (jakearchibald) [reply src/lib/sw-bridge.ts]: fwiw, `BroadcastChannel` is only used where Share Target V2 is also supported, but I'll make it work using the clients API anyway.

**Comment by jakearchibald** [comment]
@surma finally removed BroadcastChannel. I've tested this, and it works. Wanna give it a review?

File: src/components/App/index.tsx [hunk 2]
```diff
@@ -103,15 +114,18 @@ export default class App extends Component<Props, State> {
...
     return (
       <div id="app" class={style.app}>
         <file-drop accept="image/*" onfiledrop={this.onFileDrop} class={style.drop}>
-          {!isEditorOpen
-            ? <Intro onFile={this.onIntroPickFile} showSnack={this.showSnack} />
-            : (Compress)
-              ? <Compress file={file!} showSnack={this.showSnack} onBack={back} />
-              : <loading-spinner class={style.appLoader}/>
+          {
+            awaitingShareTarget
```
Comment (surma) [inline src/components/App/index.tsx]: This ternary chain is getting a bit outta hand, don’t you think? Maybe we can put it into some helper functions and a switch? `shouldShowSpinner(state)` `shouldShowIntro(state)` etc?

File: src/sw/util.ts [hunk 3]
```diff
@@ -104,3 +124,26 @@ export async function cacheAdditionalProcessors(cacheName: string, buildAssets:
...
+function nextMessage(dataVal: string): Promise<void> {
+  return new Promise((resolve) => {
+    if (!nextMessageResolveMap.has(dataVal)) {
+      nextMessageResolveMap.set(dataVal, []);
+    }
+    nextMessageResolveMap.get(dataVal)!.push(resolve);
+  });
+}
+
+self.addEventListener('message', (event) => {
```
Comment (surma) [inline src/sw/util.ts]: Curious: Why the indirection via a map? Couldn’t each call to `nextMessage()` just add an additional listener that self-removes when a message with the desired string has been received?

**Comment by surma** [comment]
Also seems like a rebase is needed

File: src/sw/util.ts (hunk 3 above)
Comment (jakearchibald) [reply src/sw/util.ts]: Top level listeners can only be added during initial execution.

**Comment by jakearchibald** [comment]
Hm, I merged master as part of this. I'll double check.

**Pull request merged**